File: DataEngineering/02_Data_Modeling/projects/golang/02_snowflake_schema_builder/main.go
package main

import (
	"fmt"
	"log"

	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// DimCountry is the top level of the normalized location hierarchy
type DimCountry struct {
	CountryID   uint   `gorm:"primaryKey;autoIncrement"`
	CountryName string `gorm:"not null"`
}

// DimState belongs to DimCountry
type DimState struct {
	StateID   uint   `gorm:"primaryKey;autoIncrement"`
	StateName string `gorm:"not null"`
	CountryID uint   `gorm:"not null"`
	Country   DimCountry `gorm:"foreignKey:CountryID"`
}

// DimCity belongs to DimState
type DimCity struct {
	CityID   uint   `gorm:"primaryKey;autoIncrement"`
	CityName string `gorm:"not null"`
	StateID  uint   `gorm:"not null"`
	State    DimState `gorm:"foreignKey:StateID"`
}

// DimStore belongs to DimCity
type DimStore struct {
	StoreID   uint   `gorm:"primaryKey;autoIncrement"`
	StoreName string `gorm:"not null"`
	CityID    uint   `gorm:"not null"`
	City      DimCity `gorm:"foreignKey:CityID"`
}

// FactSales links to the most granular dimension
type FactSales struct {
	SaleID  uint    `gorm:"primaryKey;autoIncrement"`
	StoreID uint    `gorm:"not null"`
	Amount  float64 `gorm:"not null"`
	Store   DimStore `gorm:"foreignKey:StoreID"`
}

func main() {
	// Standard connection to a relational database for schema management.
	dsn := "host=localhost user=postgres password=postgres dbname=data_warehouse port=5432 sslmode=disable"
	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Info),
	})
	if err != nil {
		log.Printf("Failed to connect to database: %v", err)
		log.Println("Skipping actual execution...")
		return
	}

	// 1. Auto Migrate Snowflake Schema (creates tables with foreign keys)
	err = db.AutoMigrate(&DimCountry{}, &DimState{}, &DimCity{}, &DimStore{}, &FactSales{})
	if err != nil {
		log.Fatalf("Error creating Snowflake schema: %v", err)
	}
	fmt.Println("Snowflake Schema successfully created/migrated!")

	// 2. Insert Normalized Data
	country := DimCountry{CountryName: "USA"}
	db.Create(&country)

	state := DimState{StateName: "New York", CountryID: country.CountryID}
	db.Create(&state)

	city := DimCity{CityName: "NYC", StateID: state.StateID}
	db.Create(&city)

	store := DimStore{StoreName: "NYC Flagship", CityID: city.CityID}
	db.Create(&store)

	sale := FactSales{StoreID: store.StoreID, Amount: 500.0}
	db.Create(&sale)
	fmt.Println("Snowflake data inserted.")

	// 3. Querying across the snowflake relationships
	type SnowflakeResult struct {
		SaleID      uint
		StoreName   string
		CityName    string
		StateName   string
		CountryName string
		Amount      float64
	}

	var results []SnowflakeResult
	db.Table("fact_sales").
		Select(`
			fact_sales.sale_id, 
			dim_stores.store_name, 
			dim_cities.city_name, 
			dim_states.state_name, 
			dim_countries.country_name, 
			fact_sales.amount
		`).
		Joins("JOIN dim_stores ON dim_stores.store_id = fact_sales.store_id").
		Joins("JOIN dim_cities ON dim_cities.city_id = dim_stores.city_id").
		Joins("JOIN dim_states ON dim_states.state_id = dim_cities.state_id").
		Joins("JOIN dim_countries ON dim_countries.country_id = dim_states.country_id").
		Scan(&results)

	fmt.Println("--- Snowflake Query Results ---")
	for _, r := range results {
		fmt.Printf("Sale: %d | Store: %s | Location: %s, %s, %s | Amount: $%.2f\n", 
			r.SaleID, r.StoreName, r.CityName, r.StateName, r.CountryName, r.Amount)
	}
}

File: DataEngineering/02_Data_Modeling/projects/python/02_snowflake_schema_builder/main.py
import logging
from sqlalchemy import create_engine, Column, Integer, String, Float, ForeignKey
from sqlalchemy.orm import declarative_base, sessionmaker, relationship

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

Base = declarative_base()

class DimCountry(Base):
    __tablename__ = 'dim_country'
    country_id = Column(Integer, primary_key=True, autoincrement=True)
    country_name = Column(String, nullable=False)

class DimState(Base):
    __tablename__ = 'dim_state'
    state_id = Column(Integer, primary_key=True, autoincrement=True)
    state_name = Column(String, nullable=False)
    country_id = Column(Integer, ForeignKey('dim_country.country_id'), nullable=False)
    
    country = relationship("DimCountry")

class DimCity(Base):
    __tablename__ = 'dim_city'
    city_id = Column(Integer, primary_key=True, autoincrement=True)
    city_name = Column(String, nullable=False)
    state_id = Column(Integer, ForeignKey('dim_state.state_id'), nullable=False)

    state = relationship("DimState")

class DimStore(Base):
    __tablename__ = 'dim_store'
    store_id = Column(Integer, primary_key=True, autoincrement=True)
    store_name = Column(String, nullable=False)
    city_id = Column(Integer, ForeignKey('dim_city.city_id'), nullable=False)

    city = relationship("DimCity")

class FactSales(Base):
    __tablename__ = 'fact_sales'
    sale_id = Column(Integer, primary_key=True, autoincrement=True)
    store_id = Column(Integer, ForeignKey('dim_store.store_id'), nullable=False)
    amount = Column(Float, nullable=False)

    store = relationship("DimStore")

def main():
    # Production configurations would use a robust RDBMS like PostgreSQL or Snowflake via SQLAlchemy dialects.
    DATABASE_URL = "sqlite:///:memory:"
    engine = create_engine(DATABASE_URL, echo=False)
    
    Base.metadata.create_all(engine)
    
    Session = sessionmaker(bind=engine)
    session = Session()

    try:
        # Hierarchical insertion for Snowflake Schema
        country = DimCountry(country_name="USA")
        session.add(country)
        session.flush()

        state = DimState(state_name="New York", country_id=country.country_id)
        session.add(state)
        session.flush()

        city = DimCity(city_name="NYC", state_id=state.state_id)
        session.add(city)
        session.flush()

        store = DimStore(store_name="NYC Flagship", city_id=city.city_id)
        session.add(store)
        session.flush()

        sale = FactSales(store_id=store.store_id, amount=500.0)
        session.add(sale)
        
        session.commit()
        logger.info("Snowflake Schema Data Generated.")

        # Query across snowflake relationships
        logger.info("Querying across the snowflake relationships:")
        results = session.query(
            FactSales.sale_id,
            DimStore.store_name,
            DimCity.city_name,
            DimState.state_name,
            DimCountry.country_name,
            FactSales.amount
        ).join(DimStore).join(DimCity).join(DimState).join(DimCountry).all()

        for r in results:
            logger.info(f"Sale: {r.sale_id} | Store: {r.store_name} | Location: {r.city_name}, {r.state_name}, {r.country_name} | Amt: ${r.amount}")

    except Exception as e:
        session.rollback()
        logger.error(f"Transaction failed: {e}")
    finally:
        session.close()

if __name__ == '__main__':
    main()
